subsys/bindesc/CMakeLists.txt

```
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
+
+if(CONFIG_X86)
+  zephyr_linker_sources(ROM_START SORT_KEY 0x1bindesc bindesc_no_vt.ld)
+  zephyr_library_sources(x86/bindesc_skip.S)
+else()
+  zephyr_linker_sources(ROM_START SORT_KEY 0x1bindesc bindesc.ld)
+endif()
+
+macro(gen_build_time_int_definition def_name format)
+  if(CONFIG_BINDESC_${def_name})
+    string(TIMESTAMP ${def_name} ${format})
+    # remove leading zeros so that the output will not be interpreted as octal
+    math(EXPR ${def_name} ${${def_name}})
+    zephyr_library_compile_definitions(${def_name}=${${def_name}})
+  endif()
+endmacro()
+
+macro(gen_build_time_str_definition def_name format)
+  if(CONFIG_BINDESC_${def_name})
+    string(TIMESTAMP ${def_name} ${${format}})
+    zephyr_library_compile_definitions(${def_name}="${${def_name}}")
+  endif()
+endmacro()
+
+macro(gen_str_definition def_name value)
+  if(CONFIG_BINDESC_${def_name})
+    zephyr_library_compile_definitions(${def_name}="${value}")
+  endif()
+endmacro()
+
+if(CONFIG_BINDESC_DEFINE_BUILD_TIME)
+  zephyr_library_sources(bindesc_build_time.c)
+  gen_build_time_int_definition(BUILD_TIME_YEAR "%Y")
+  gen_build_time_int_definition(BUILD_TIME_MONTH "%m")
+  gen_build_time_int_definition(BUILD_TIME_DAY "%d")
+  gen_build_time_int_definition(BUILD_TIME_HOUR "%H")
+  gen_build_time_int_definition(BUILD_TIME_MINUTE "%M")
+  gen_build_time_int_definition(BUILD_TIME_SECOND "%S")
+  gen_build_time_int_definition(BUILD_TIME_UNIX "%s")
+
+  gen_build_time_str_definition(BUILD_DATE_TIME_STRING
+      CONFIG_BINDESC_BUILD_DATE_TIME_STRING_FORMAT)
+  gen_build_time_str_definition(BUILD_DATE_STRING
+      CONFIG_BINDESC_BUILD_DATE_STRING_FORMAT)
+  gen_build_time_str_definition(BUILD_TIME_STRING
+      CONFIG_BINDESC_BUILD_TIME_STRING_FORMAT)
+
+  if(CONFIG_BINDESC_BUILD_TIME_ALWAYS_REBUILD)
+    # By adding a custom target that tries to unset a dummy variable,
+    # CMake is forced to rebuild this target on every build. This is
+    # done to ensure that the timestamp is always up to date.
+    add_custom_target(
+      bindesc_time_force_rebuild
+      COMMAND ${CMAKE_COMMAND} -U BUILD_TIME_DUMMY ${CMAKE_BINARY_DIR}
```

follow up to comment here: https://github.com/zephyrproject-rtos/zephyr/pull/54464#discussion_r1337379208

if it is purely a matter of re-invoking CMake, then there is no need for the `-U`, and this proposal should be sufficient:
```suggestion
      COMMAND ${CMAKE_COMMAND} ${CMAKE_BINARY_DIR}
```

that said, then i'm generally not fond of build system implementation where the following cannot be achieved when there are no changes to the system:
```
$ ninja
ninja: no work to do.
```
but in this case where the behavior is guarded by `CONFIG_BINDESC_BUILD_TIME_ALWAYS_REBUILD` then it is accepted.